### PR TITLE
[FLINK-31980] Implement support for EFO in Kinesis consumer

### PR DESCRIPTION
**Pull request by hlteoh37**
## Purpose of the change

Implement support for EFO in the new KDS source

## Verifying this change

This change added tests and can be verified as follows:
- Added unit tests
- Manually verified by running the Kinesis connector on a local Flink cluster.

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [x] New feature has been introduced
  - If yes, how is this documented? Documented as part of https://issues.apache.org/jira/browse/FLINK-31989

### CODE REVIEW

**Comment by Sunninsky** [comment]
Hi, EFO is very important for using the New SouceAPI. Do you plan to finish this PR?

**Comment by hlteoh37** [comment]
@Sunninsky Hey! We are currently working on implementing EFO support for the new source connector. Estimate maybe end of the quarter?

**Comment by Sunninsky** [comment]
> @Sunninsky Hey! We are currently working on implementing EFO support for the new source connector. Estimate maybe end of the quarter?

That's very cool, thank you.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java [hunk 1]
```diff
@@ -163,9 +180,51 @@ public KinesisStreamsSourceEnumeratorState snapshotState(long checkpointId) thro
 
     @Override
     public void close() throws IOException {
+        efoConsumerCleanup(sourceConfig, streamConsumerRegistrar);
         streamProxy.close();
     }
 
+    private void efoConsumerStartSetup(
```
Comment (dzikosc) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:187]: It appears that this code is fairly independent from the rest of this class. As per single responsibility principle, can we move this code away to a dedicated class and then think about the best way to plug it into the Enumerator lifecycle?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java [hunk 2]
```diff
@@ -57,4 +69,20 @@ public enum InitialPosition {
...
+
+    public static final ConfigOption<ConsumerLifecycle> EFO_CONSUMER_LIFECYCLE =
+            ConfigOptions.key("efo.consumer.lifecycle")
+                    .enumType(ConsumerLifecycle.class)
+                    .defaultValue(ConsumerLifecycle.JOB_MANAGED)
+                    .withDescription(
+                            "Setting to control whether the lifecycle of EFO consumer is managed by the Flink job. If JOB_MANAGED, then the Flink job will register the consumer on startup and deregister it on shutdown.");
+
+    public static final ConfigOption<String> EFO_CONSUMER_NAME =
+            ConfigOptions.key("efo.consumer.name").stringType().noDefaultValue();
```
Comment (guanlin84) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java:87]: Propose to use `kinesis.stream.efo.consumer.name` as the config key here. As I will be updating the public example to use this.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java [hunk 3]
```diff
@@ -57,4 +69,20 @@ public enum InitialPosition {
                     .longType()
                     .defaultValue(10000L)
                     .withDescription("The interval between each attempt to discover new shards.");
+
+    public static final ConfigOption<ReaderType> READER_TYPE =
+            ConfigOptions.key("type")
```
[the remark this answers is not in the document]
Comment (guanlin84) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java:74]: Propose to use `kinesis.stream.reader.type` as the config key. As I will be updating the public example to use this.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java [hunk 4]
```diff
@@ -57,4 +69,20 @@ public enum InitialPosition {
...
+                    .enumType(ReaderType.class)
+                    .defaultValue(ReaderType.POLLING)
+                    .withDescription("The type of reader used to read from the Kinesis stream.");
+
+    public static final ConfigOption<ConsumerLifecycle> EFO_CONSUMER_LIFECYCLE =
+            ConfigOptions.key("efo.consumer.lifecycle")
```
Comment (guanlin84) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java:80]: Propose to use `kinesis.stream.efo.consumer.lifecycle` as the config key. As I will be updating the public example to use this.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java [hunk 5]
```diff
@@ -32,6 +32,18 @@ public enum InitialPosition {
...
+    }
+
+    /** Defines lifecycle management of EFO consumer on Kinesis stream. */
+    public enum ConsumerLifecycle {
+        JOB_MANAGED,
+        SELF_MANAGED
+    }
+
     public static final ConfigOption<InitialPosition> STREAM_INITIAL_POSITION =
             ConfigOptions.key("flink.stream.initpos")
```
Comment (guanlin84) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java:48]: Shouldn't this be `kinesis.stream.init.position` ?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSource.java [hunk 6]
```diff
@@ -178,6 +185,84 @@ public SimpleVersionedSerializer<KinesisShardSplit> getSplitSerializer() {
...
+            case POLLING:
+                return getPollingKinesisShardSplitReaderSupplier(sourceConfig, shardMetricGroupMap);
+            case EFO:
+                return getFanOutKinesisShardSplitReaderSupplier(sourceConfig, shardMetricGroupMap);
+            default:
+                throw new IllegalArgumentException("Unsupported reader type: " + readerType);
+        }
+    }
+
+    private Supplier<SplitReader<Record, KinesisShardSplit>>
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSource.java:223]: Done

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSource.java [hunk 7]
```diff
@@ -178,6 +185,84 @@ public SimpleVersionedSerializer<KinesisShardSplit> getSplitSerializer() {
...
+                    "Unable to lookup consumer ARN for stream "
+                            + streamArn
+                            + " and consumer "
+                            + consumerName,
+                    e);
+        }
+    }
+
+    private KinesisAsyncStreamProxy createKinesisAsyncStreamProxy(Configuration consumerConfig) {
+        SdkAsyncHttpClient asyncHttpClient =
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSource.java:274]: Rejiggled to make this close to the end

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java (hunk 1 above)
Comment (hlteoh37) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:187]: Refactored logic into existing `StreamConsumerRegistrar`

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/reader/fanout/StreamConsumerRegistrar.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.kinesis.model.RegisterStreamConsumerResponse;
+
+/** Responsible for registering and deregistering EFO stream consumers. */
+@Internal
+public class StreamConsumerRegistrar {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamConsumerRegistrar.class);
+
+    private final StreamProxy kinesisStreamProxy;
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/reader/fanout/StreamConsumerRegistrar.java:33]: I think it fits better that the proxy simply wraps the client. It's more extensible for the future if we want to include Async as well.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/split/StartingPosition.java [hunk 9]
```diff
@@ -52,6 +53,30 @@ public Object getStartingMarker() {
         return startingMarker;
     }
 
+    public software.amazon.awssdk.services.kinesis.model.StartingPosition getSdkStartingPosition() {
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/split/StartingPosition.java:56]: Refactored into Util

**Pull request merged**